### ISSUE TITLE
Model output (x,y,w,h,r) no post-processed during validation vs. inference

### ISSUE THREAD

**Issue by Petros626** [ultralytics/ultralytics#20202]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hey,

Did I get it right, that during validation (`model.val()`) the rotated box points (xy1, xy2, xy3, xy4) are directly created by the model head? The raw angle output is -45°-135°.

In opposite to that using `model.predict()` requires post-processing ([0°,90°]) function. Why is that so?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Petros626, thank you for reaching out and for your detailed question about model outputs in Ultralytics 🚀! This is an automated response to help get you the fastest support, and an Ultralytics engineer will also assist you here soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/) where you’ll find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, which may already address your question.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate.

For custom training or usage ❓ Questions, please include as much detail as possible, like code snippets, sample data, and logs. Also, make sure to follow our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join our community wherever you prefer:
- Real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions on [Discourse](https://community.ultralytics.com/)
- Share and learn on our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Please ensure you're using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

You can run YOLO in any of these up-to-date, verified environments (all dependencies and [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience and for helping improve Ultralytics!

**Comment by glenn-jocher** [comment]
You're correct. During validation (`model.val()`), the model's raw output angle range is indeed -45° to 135° (or [-π/4, 3π/4] in radians) as shown in the model head code:

```python
angle = (angle.sigmoid() - 0.25) * math.pi  # [-pi/4, 3pi/4]
```

For prediction (`model.predict()`), there's an additional post-processing step that regularizes the angles to [0°, 90°] range using `regularize_rboxes()`:

```python
rboxes = ops.regularize_rboxes(torch.cat([pred[:, :4], pred[:, -1:]], dim=-1))
```

This difference exists because:
1. During validation, we evaluate the raw model outputs against ground truth
2. During inference, we want to provide users with standardized outputs in a consistent format

The regularization simplifies downstream processing by ensuring all rotated boxes have angles in a standard range, making the outputs easier to work with in applications.

**Comment by Petros626** [comment]
There were attempts using the raw output (-45°-135°)  on a prediction set, like DOTA? So that you can see a difference in performance

**Comment by glenn-jocher** [comment]
I'm not aware of any published performance comparisons specifically testing raw angle output (-45° to 135°) versus regularized angles (0° to 90°) on the DOTA dataset. The angle regularization step in `ops.regularize_rboxes()` primarily helps with consistent representation rather than changing detection accuracy.

The regularization works by swapping width and height when needed and normalizing the angle range, which is mathematically equivalent to the original representation. From the code in `regularize_rboxes()`:

```python
swap = t % math.pi >= math.pi / 2
w_ = torch.where(swap, h, w)
h_ = torch.where(swap, w, h)
t = t % (math.pi / 2)
```

This standardization makes post-processing and visualization more consistent but shouldn't impact the actual detection performance since it's a lossless transformation.
